lucene/core/src/test/org/apache/lucene/index/TestTieredMergePolicy.java

```diff
@@ -77,20 +77,23 @@ protected void assertSegmentInfos(MergePolicy policy, SegmentInfos infos) throws
     long levelSizeBytes = Math.max(minSegmentBytes, (long) (tmp.getFloorSegmentMB() * 1024 * 1024));
     long bytesLeft = totalBytes;
     double allowedSegCount = 0;
+    final int maxNumSegmentsOnHighestTier =
+        (int) Math.max(tmp.getSegmentsPerTier(), tmp.getTargetSearchConcurrency());
     // below we make the assumption that segments that reached the max segment
     // size divided by 2 don't need merging anymore
     int mergeFactor = (int) Math.min(tmp.getSegmentsPerTier(), tmp.getMaxMergeAtOnce());
     while (true) {
       final double segCountLevel = bytesLeft / (double) levelSizeBytes;
-      if (segCountLevel < tmp.getSegmentsPerTier() || levelSizeBytes >= maxMergedSegmentBytes / 2) {
+      if (segCountLevel < maxNumSegmentsOnHighestTier
```

Comment:
Should it allow `maxNumSegmentsOnHighestTier` for consistency with the code under `main`? Though I don't expect it to make a big difference since `segCountLevel` is a double.

```suggestion
      if (segCountLevel <= maxNumSegmentsOnHighestTier
```